accounting/invoice/models.py

```diff
@@ -20,3 +20,376 @@
+    auto_upload_google_drive_on_save = models.BooleanField(  # pylint: disable=invalid-name
+        default=False,
+        help_text=_("Whether this invoice should be uploaded to Google Drive when converted to a PDF on save."))
+
+    class Meta:
+        verbose_name = _('Invoice')
+        verbose_name_plural = _('Invoices')
+
+    def __str__(self):
+        """
+        Indicate between who this invoice is, and whether it has been paid.
+        """
+        return '{date}: {provider} invoicing {client} ({paid})'.format(
+            date=self.date,
+            provider=self.provider,
+            client=self.client,
+            paid='PAID' if self.paid else 'PENDING',
+        )
+
+    @property
+    def hourly_rate(self):
+        """
+        Get the `HourlyRate` object between the provider and client of this invoice.
+        """
+        return self.provider.provider_hourly_rates.get(client=self.client)
+
+    def aggregate_line_items(self, fields=None):
+        """
+        Aggregate this invoice's line items so quantities and total costs of items with the same keys are summed.
+
+        :param fields: Which fields should be returned for each aggregated line item.
+        :return: A `QuerySet` of `dict`s representing each aggregated line item.
+        """
+        if fields is None:
+            fields = ('key', 'name', 'price', 'quantity', 'total',)
+
+        # This returns line items with distinct keys, names, and prices, with their quantities and totals
+        # summed up into single `quantity` and `total` fields. With this, the PDF will have a single line
+        # for each distinct key. For example, each line item can correspond to a single JIRA task, with
+        # all worklogs summed up.
+        return (self.line_items
+                .values('key')
+                .distinct()
+                .annotate(
+                    total=models.Sum(models.F('quantity') * models.F('price'), output_field=models.DecimalField()),
+                    quantity=models.Sum('quantity'))
+                .values(*fields)
+                .order_by('-key'))
+
+    def fill_line_items_from_jira(self):
+        """
+        An idempotent way to fill this Invoice's line items with JIRA worklogs.
+
+        Each line item created through this method is tagged to be recognized as having come through JIRA. This
+        allows us to only update those line items through this function that previously came from JIRA.
+
+        Does not work if JIRA integration is disabled.
+
+        TODO: This should be a synchronous job on a non-web worker.
+        """
+        if not settings.ENABLE_JIRA:
+            return
+
+        jira = Jira()
+        jira_worklogs = jira.tempo_worklogs(self.provider.user.username, self.billing_start_date, self.billing_end_date)
+        jira_worklog_set = {
+            (worklog.worklog_id, worklog.issue_key, worklog.issue_title, worklog.description, worklog.time_spent,)
```

Comment:
I was thinking of doing this, yep. Will do it now then.

Reply:
Aw darn, this is not easy to do because later below I find the difference between this set which could contain namedtuples, and the QuerySet of existing line items which is not of namedtuples, and this type difference causes a problem. It seems https://github.com/django/django/pull/8852 was merged but isn't in the version we're using; I assume it's in 2.0 but I can't use 2.0 yet because some dependencies don't work with it. I wish I had time to upstream updates for them.